### DISCUSSION TITLE
Best Practises Check-in

### DISCUSSION THREAD

**Discussion by Sharpz7** [coder/code-server#5959] [Q&A] [answered]
Hi,

Currently I use code-server with a custom 5GB+ custom image, where I manually install everything through a mix of `apt` and custom installs, including docker (for use through the docker sock), docker-compose and a bunch of different programming languages. I then control this from docker-compose.

My plan is to move to k8s, and also use DinD, as not having localhost is starting to become more and more problematic.

I was hoping people could give me advice on best practises for building docker images, volume mounting and accessing docker when I do this move, for example:

- I need to move between 3+ languages / day, should I have multiple images I switch between, and should I have multiple instances of code-server on my server ready to switch between them, that all access the same file system?
- Any recommendaitions for getting DinD and code-server working together? 

Also, if people think I am missing something, please let me know. I know coder exists, but, correct me if I am wrong, I will still need to manage my own images if I want to have the languages and different components pre-installed? There's no equivalent to `devcontainer.json` like what github codespaces has? Or, does coder have some other method for adding more functionality to code-server without a custom docker image?

Thanks in advance :))

**Comment by jsjoeio** [comment]
@bpmct has worked with a lot of different teams so he might have a suggestion or two!

**Comment by bpmct** [comment, marked as answer]
> Any recommendations for getting DinD and code-server working together?

I'd actually recommend running the Sysbox container runtime on Kubernetes and installing code-server and Docker into a base container. It has the best isolation and you can use more Docker features.

> I need to move between 3+ languages / day, should I have multiple images I switch between, and should I have multiple instances of code-server on my server ready to switch between them, that all access the same file system?

We typically see teams build one giant, base image with all the dependencies. However, you could also mount the same volume in multiple containers in a pod if you want some process isolation for the services but needed to share the files.

> Or, does coder have some other method for adding more functionality to code-server without a custom docker image?

devcontainer support is coming! coder/coder#5559. There are ways to get devcontainers working with Coder today, but it's definitely a workaround. The example I have uses a VM with nested Docker, not dind. However, that would also be possible.

**Reply by Sharpz7** [reply]
Thanks! This looks great. 
> I'd actually recommend running the Sysbox container runtime on Kubernetes and installing code-server and Docker into a base container. It has the best isolation and you can use more Docker features.

My experience with sysbox is limited, do you have advice on where to get a more step-by-step guide to setting this up with code-server and potentially coder as well?

> devcontainer support is coming! https://github.com/coder/coder/issues/5559. There are ways to get devcontainers working with Coder today, but it's definitely a workaround. The example I have uses a VM with nested Docker, not dind. However, that would also be possible.

I never knew about the devcontainers cli - I assumed it was closed source. This is greatly appreciated :)) Again, if you have a larger, more basic example usage of this that would also be greatly appreciated.
